File: src/lib/astraura/experiencias.ts
// Experiencias del razonador (Trinidad: Needle · Jev · BitNet) en el OS.
// Mismo formato que scripts/puente/experiencias.py del Mando. Se guardan en
// IndexedDB (nunca localStorage: son muchas) detrás de la interfaz Almacen.

export type Capa = "regla" | "needle" | "jev" | "bitnet" | "llm" | "persona";
export type TipoExperiencia = "intencion" | "eleccion" | "si_no" | "puntuacion" | "texto";

export const MAX_ENTRADA = 400;

export interface Experiencia {
  id: string;
  t: string;
  medio: string;
  capa: Capa;
  tipo: TipoExperiencia;
  dominio: string;
  entrada: string;
  salida: unknown;
  confianza: number | null;
  ms: number | null;
  resultado: boolean | null;
  opciones?: unknown[];
  nota_resultado?: string;
}

export interface CierreExperiencia {
  ref: string;
  t: string;
  resultado: boolean;
  nota: string;
}

// Clave-valor mínimo: IndexedDB en el navegador, un Map en las pruebas.
export interface Almacen {
  poner(linea: Experiencia | CierreExperiencia): Promise<void>;
  lineas(): Promise<(Experiencia | CierreExperiencia)[]>;
}

export function recortar(x: unknown, n = MAX_ENTRADA): string {
  const s = typeof x === "string" ? x : JSON.stringify(x) ?? "";
  return s.length <= n ? s : s.slice(0, n) + "…";
}

function ahora(): string {
  return new Date().toISOString().replace("T", " ").slice(0, 19);
}

async function idDe(semilla: string): Promise<string> {
  const datos = new TextEncoder().encode(semilla);
  const cripto = globalThis.crypto?.subtle;
  if (cripto) {
    const hash = await cripto.digest("SHA-256", datos);
    return Array.from(new Uint8Array(hash)).map((b) => b.toString(16).padStart(2, "0")).join("").slice(0, 12);
  }
  let h = 0;
  for (const b of datos) h = (h * 31 + b) >>> 0;
  return h.toString(16).padStart(8, "0") + "0000";
}

export async function nueva(op: {
  capa: Capa;
  tipo: TipoExperiencia;
  entrada: unknown;
  salida: unknown;
  confianza?: number | null;
  ms?: number | null;
  opciones?: unknown[];
  dominio?: string;
  medio?: string;
}): Promise<Experiencia> {
  const entrada = recortar(op.entrada);
  const e: Experiencia = {
    id: await idDe(`${Date.now()}|${op.capa}|${op.tipo}|${entrada}`),
    t: ahora(),
    medio: op.medio ?? "os-web",
    capa: op.capa,
    tipo: op.tipo,
    dominio: op.dominio ?? "",
    entrada,
    salida: op.salida,
    confianza: op.confianza == null ? null : Math.round(op.confianza * 10000) / 10000,
    ms: op.ms == null ? null : Math.trunc(op.ms),
    resultado: null,
  };
  if (op.opciones && op.opciones.length > 0) e.opciones = [...op.opciones];
  return e;
}

export async function anotar(e: Experiencia, almacen: Almacen): Promise<string> {
  await almacen.poner(e);
  return e.id;
}

export async function cerrar(id: string, resultado: boolean, nota: string, almacen: Almacen): Promise<CierreExperiencia> {
  const c: CierreExperiencia = { ref: id, t: ahora(), resultado: Boolean(resultado), nota: nota.slice(0, 200) };
  await almacen.poner(c);
  return c;
}

export async function leer(ultimas: number = 500, almacen: Almacen): Promise<Experiencia[]> {
  const lineas = (await almacen.lineas()).slice(-ultimas * 2);
  const porId = new Map<string, Experiencia>();
  const orden: string[] = [];
  for (const d of lineas) {
    if ("ref" in d) {
      const e = porId.get(d.ref);
      if (e) {
        e.resultado = d.resultado;
        e.nota_resultado = d.nota ?? "";
      }
    } else if (d.id) {
      porId.set(d.id, { ...d });
      orden.push(d.id);
    }
  }
  return orden.slice(-ultimas).map((id) => porId.get(id) as Experiencia);
}

export async function exportar(almacen: Almacen): Promise<string> {
  const lineas = await almacen.lineas();
  return lineas.map((l) => JSON.stringify(l)).join("\n") + (lineas.length > 0 ? "\n" : "");
}

File: scripts/puente/experiencias.py
"""Experiencias: cada decisión de la Trinidad (Needle · Jev · BitNet) queda anotada.

(2026-09-20) Es la materia prima de la conciencia colectiva: qué se preguntó, qué capa
contestó, con qué confianza, cuánto tardó y —cuando se sabe— si acertó. De aquí salen
los JSONL con los que `needle finetune` entrena el adaptador colectivo cada noche y las
curvas de calibración de Jev (¿cuando dice 0,8 acierta el 80 %?).

Formato: una línea JSON por experiencia en ~/.starseed/experiencias.jsonl (append-only)
y copia en starseed_memory_root/aprendizaje/experiencias/<host>.jsonl para que el espejo
de Drive y la mesh la repartan. Nunca lleva claves ni texto completo de prompts largos:
`entrada` se recorta a 400 caracteres.
"""

import hashlib
import json
import os
import socket
import time

RUTA = os.path.expanduser("~/.starseed/experiencias.jsonl")
RAIZ = os.path.abspath(
    os.path.join(os.path.dirname(os.path.abspath(__file__)), "..", "..")
)
COPIA_DIR = os.path.join(RAIZ, "starseed_memory_root", "aprendizaje", "experiencias")
CAPAS = ("regla", "needle", "jev", "bitnet", "llm", "persona")
MAX_ENTRADA = 400


def _host():
    try:
        return socket.gethostname().split(".")[0]
    except Exception:
        return "nodo"


def _recortar(x, n=MAX_ENTRADA):
    s = x if isinstance(x, str) else json.dumps(x, ensure_ascii=False, sort_keys=True)
    return s if len(s) <= n else s[:n] + "…"


def nueva(
    capa, tipo, entrada, salida, confianza=None, ms=None, opciones=None, dominio=""
):
    """Construye una experiencia (dict) sin escribirla. `tipo`: intencion | eleccion | si_no | puntuacion | texto."""
    if capa not in CAPAS:
        raise ValueError("capa desconocida: %s" % capa)
    e = {
        "id": hashlib.sha256(
            ("%s|%s|%s|%s" % (time.time(), capa, tipo, _recortar(entrada))).encode()
        ).hexdigest()[:12],
        "t": time.strftime("%Y-%m-%d %H:%M:%S"),
        "medio": _host(),
        "capa": capa,
        "tipo": tipo,
        "dominio": dominio,
        "entrada": _recortar(entrada),
        "salida": salida,
        "confianza": None if confianza is None else round(float(confianza), 4),
        "ms": None if ms is None else int(ms),
        "resultado": None,
    }
    if opciones:
        e["opciones"] = list(opciones)
    return e


def anotar(e, ruta=None):
    """Escribe la experiencia (una línea) en el registro y en la copia del memory root."""
    linea = json.dumps(e, ensure_ascii=False) + "\n"
    for destino in (ruta or RUTA, os.path.join(COPIA_DIR, _host() + ".jsonl")):
        try:
            os.makedirs(os.path.dirname(destino), exist_ok=True)
            with open(destino, "a", encoding="utf-8") as f:
                f.write(linea)
        except OSError:
            pass
    return e.get("id") or e.get("ref")


def resultado(id_, acierto, nota="", ruta=None):
    """Cierra una experiencia con lo que pasó de verdad (True/False) — una línea de referencia."""
    return anotar(
        {
            "ref": id_,
            "t": time.strftime("%Y-%m-%d %H:%M:%S"),
            "resultado": bool(acierto),
            "nota": nota[:200],
        },
        ruta,
    )


def leer(ruta=None, ultimas=500):
    """Las últimas experiencias con su resultado aplicado (si llegó)."""
    try:
        with open(ruta or RUTA, encoding="utf-8") as f:
            lineas = f.read().splitlines()[-ultimas * 2 :]
    except OSError:
        return []
    por_id, orden = {}, []
    for l in lineas:
        try:
            d = json.loads(l)
        except Exception:
            continue
        if "ref" in d:
            if d["ref"] in por_id:
                por_id[d["ref"]]["resultado"] = d.get("resultado")
                por_id[d["ref"]]["nota_resultado"] = d.get("nota", "")
        elif d.get("id"):
            por_id[d["id"]] = d
            orden.append(d["id"])
    return [por_id[i] for i in orden[-ultimas:]]


def calibracion(exps, capa="jev"):
    """{tramo: {n, aciertos}} por décimas de confianza, solo con resultado conocido."""
    tramos = {}
    for e in exps:
        if (
            e.get("capa") != capa
            or e.get("resultado") is None
            or e.get("confianza") is None
        ):
            continue
        k = "%.1f" % (int(e["confianza"] * 10) / 10)
        t = tramos.setdefault(k, {"n": 0, "aciertos": 0})
        t["n"] += 1
        t["aciertos"] += 1 if e["resultado"] else 0
    return tramos


def para_needle(exps):
    """Las experiencias de intención con acierto → líneas JSONL que entiende `needle finetune`."""
    fuera = []
    for e in exps:
        if e.get("tipo") != "intencion" or not e.get("resultado"):
            continue
        s = e.get("salida") or {}
        if (
            not isinstance(s, dict)
            or not s.get("herramientas")
            or not s.get("llamadas")
        ):
            continue
        fuera.append(
            {
                "query": e["entrada"],
                "tools": s["herramientas"],
                "answers": [
                    {"name": c["nombre"], "arguments": c.get("argumentos") or {}}
                    for c in s["llamadas"]
                ],
                "reasoning": s.get("razonamiento") or "",
            }
        )
    return fuera
